### ISSUE TITLE
Coreference resolution model performance

### ISSUE THREAD

**Issue by allanj** [allenai/allennlp#2956]
**System (please complete the following information):**
 - OS: [e.g. OSX, Linux]
 - Python version: [e.g. 3.6.1]
 - AllenNLP version: [e.g. v0.8.4, or "I installed from master"]
 - PyTorch version: (if you installed it yourself)

**Question**

* I train the coreference model with the official script [coref.jsonnet](https://github.com/allenai/allennlp/blob/master/training_config/coref.jsonnet) but only obtain F1 performance of 61.75%, which is lower than the result (63%) reported in https://allennlp.org/models. 
   ```bash 
   allennlp train coref.jsonnet -s ./tmp
   ```

   Then I also run the provided [model](https://s3-us-west-2.amazonaws.com/allennlp/models/coref-model-2018.02.05.tar.gz) on the test set and obtain 62.69%, which kind of matches the reported 63% F1. To investigate the reason, I downloaded the provided model and found the difference of configuration (i.e. json file) is `max_antecedents`. 

   So in the website https://allennlp.org/models, the `max_antecedents` was mentioned to be set as `100` but it is `150` when I unzip the provided model file and check the json file. Also, in the latest [coref.jsonnet](https://github.com/allenai/allennlp/blob/master/training_config/coref.jsonnet) script, the `max_antecedents` is `100`. 
   
   Is it the reason causing the difference? (I'm running the experiment again with 150)

* Attention calculation
  It seems there is some difference in the attention calculation compared to the original [paper](https://arxiv.org/pdf/1707.07045.pdf). 

  ![Equation](http://people.sutd.edu.sg/~allanjie/wp-content/uploads/2019/06/equation.png)

  Here the alpha_t is calculated by the contextualized representations from LSTM, and the final representation is weighted on the word representations.
  
  However, the [code](https://github.com/allenai/allennlp/blob/master/allennlp/models/coreference_resolution/coref.py#L177) in allennlp seems only consider the word representations. 
  
   Please correct me if I'm wrong.

**Comment by LoicGrobol** [comment]
The scores reported in the papers were obtained using [the original TF implementation](https://github.com/kentonl/e2e-coref), the one in AllenNLP is only a port and I don't think there has been actual performance assessments.
For the head-finding attention, the original implementation [used the LSTM hidden states](https://github.com/kentonl/e2e-coref/blob/22085000e37b007362a6577b1ba594125907bf06/coref_model.py#L350) [for the scores and the values both](https://github.com/kentonl/e2e-coref/blob/22085000e37b007362a6577b1ba594125907bf06/coref_model.py#L234) but maybe it was skipped in the port.

**Comment by allanj** [comment]
> The scores reported in the papers were obtained using [the original TF implementation](https://github.com/kentonl/e2e-coref), the one in AllenNLP is only a port and ~I don't think there has been actual performance assessments~ my bad, it's claimed there has been.
> For the head-finding attention, the original implementation [used the LSTM hidden states](https://github.com/kentonl/e2e-coref/blob/22085000e37b007362a6577b1ba594125907bf06/coref_model.py#L350) [for the scores and the values both](https://github.com/kentonl/e2e-coref/blob/22085000e37b007362a6577b1ba594125907bf06/coref_model.py#L234) but maybe it was skipped in the port.
> 
> (Disclaimer: I am the author of neither :-) )

Not so sure what do you by "port"?

An update: I change the `max_antecedents`  to `150` and modify the head-finding attention, obtain 62.66% F1. 

Also running an experiment for just changing the `max_antecedents`  to `150` without modifying the attention calculation.

**Comment by Evpok** [comment]
> Not so sure what do you by "port"?

[That the original was in TF and this is in AllenNLP/Pytorch](https://en.wikipedia.org/wiki/Porting)

**Comment by LoicGrobol** [comment]
> Not so sure what do you by "port"?

[That the original was in TF and this is in AllenNLP/Pytorch](https://en.wikipedia.org/wiki/Porting)

**Comment by DeNeutoy** [comment]
Ok so I think the difference is close enough to the result we reported that this is just a random seed issue - we should update the coref.jsonnet config to have max_anteccedents == 150 though.

**Comment by allanj** [comment]
BTW, we also implemented this AllenNLP version using BERT as well, which achieves 73.8% F1 with the best variant. Are we going to update the model for the demo?

**Comment by DeNeutoy** [comment]
Oh sweet - yes that would be great. @pdasigi can you oversee the integration of that BERT model into the demo?

**Comment by pdasigi** [comment]
@DeNeutoy Sure! I'll ping you when I have some questions though. I haven't done anything related to demos, but I'd like to learn.

**Comment by DeNeutoy** [comment]
So long as the new model is essentially the same as the previous one, should be as simple as this: 

https://github.com/allenai/allennlp-demo/pull/214

**Comment by DeNeutoy** [comment]
But sure, lmk if you have issues.
